### ISSUE TITLE
RTL control bar is reversed 

### ISSUE THREAD

**Issue by GreenLunar** [exaile/exaile#144]
No physical DVD/Video player has its controllers set from right-to-left.

FORCE LTR WIDGET DIRECTION.

See:
http://sourceforge.net/p/smplayer/bugs/584/
https://git.gnome.org/browse/ekiga/commit/?id=76c9b23
https://trac.gajim.org/changeset/5f8bfea2054a

**Comment by sjohannes** [comment]
A comment in the first link suggests that physical DVD players with reversed controls also exist and are more popular in some regions:

> Sorry, I think reversing the control bar for RTL UIs is the correct behavior, physical DVD players were made in countries that speak LTR languages, RTL (at least Arabic) users expect the "next" in a video/audio to be on the left while the "previous" is on the right. Many major software media players implement this behavior, including Windows Media Player, Totem on Linux... In fact, I was going to report this change as a bug in the latest version. Also, the icons of "Play", "Previous", "Next", "Fast forward", "Rewind" should all be mirrored.

I don't have enough information to decide either way, but I don't mind if an option is added for this.

**Comment by genodeftest** [comment]
As I read from the links it seems like the international cross-platform consensus seems to be a inconsistent behavior mixing RTL and LTR in RTL languages. I don't quite get the reason why, but if you say so…

Can you please be more specific what is wrong? Do you want the play/prevous/next icons mirrored (rtl)? Do you want the volume slider and the progress bar mirrored (rtl)? Because for me when setting `LC_ALL=ar` or `LC_ALL=he` the progress bar, volume level and play/pause button controls are left ltr and I don't even know why. Only with GtkInspector (only current exaile-gi development branch since it requires Gtk3) used to set "Visual" → "Text Direction" to "Right-to-left" this changes and everything starts right. Weirdly the behavior of the "play" icon is still inconsistent. Sometimes it points right (in the playlist and its tab title) and sometimes it points left (in the play/pause button).

In short: Which direction do you expect for these elements?
* seek bar
* volume slider
* play button/icon
* previous, next buttons/icons
* did I miss anything?
If yes, is this true in general? Because then it should be a bug in Gtk+.

**Comment by GreenLunar** [comment]
> A comment in the first link suggests that physical DVD players with reversed controls also exist and are more popular in some regions:

I have not yet seen any proof to that; please note that Windows Media Player in RTL also has its controllers in LTR direction.  Also, run `LANG=he_IL.utf8 exaile`, seek using the seek bar, and see how buggy the seek bar is.

> In short: Which direction do you expect for these elements?

Exactly as it is in English Exaile; previous, play, stop, next, seek bar.

This is the expected behavior for reversed (RTL) user interface in GTK+ and Qt, too, though there is a bug in PyGTK with the mirror effect while seeking, which does not occur with GTK+ media player that are written in other programing languages, such as C++.

Sometimes, deviating from procedures (e.g. forcing direction) is needed, like in media player controllers, keypad and other controllers that are, in reality, never conveyed in another direction.

**Comment by GreenLunar** [comment]
Another dial pad with forced direction https://trac.gajim.org/changeset/5f8bfea2054a

**Comment by GreenLunar** [comment]
> In short: Which direction do you expect for these elements?

Only the controllers to be forced to LTR, the rest may remain as they are.
Exactly like the English Exaile, force this order: previous, play, stop, next, seek bar.

**Comment by GreenLunar** [comment]
BUMP!

**Comment by sjohannes** [comment]
I'm going to mark this "patch welcome", because I don't think Dustin or I have the knowledge or motivation to create a config option for this.

**Comment by GreenLunar** [comment]
In no way, this is a configuration issue; this is a UI problem, one with the buggy mirror effect of PyGTK and the other that a reversed direction (right-to-left) is false.

Direction of media controls must always be from left to right.  It is that simple.

**Comment by GreenLunar** [comment]
[sjohannes](#issuecomment-144968106) wrote:

> A comment in the first link suggests that physical DVD players with reversed controls also exist and are more popular in some regions:

> > Sorry, I think reversing the control bar for RTL UIs is the correct behavior, physical DVD players were made in countries that speak LTR languages, RTL (at least Arabic) users expect the "next" in a video/audio to be on the left while the "previous" is on the right. Many major software media players implement this behavior, including Windows Media Player, Totem on Linux... In fact, I was going to report this change as a bug in the latest version. Also, the icons of "Play", "Previous", "Next", "Fast forward", "Rewind" should all be mirrored.

I have two comments on that:

1. The people who have asked FreeDesktop.org to include *rtl.svg icons even for media controllers are most likely NOT programmers and just kiddish enthusiasts who are impressed by themselves as to how on earth they managed to install Linux on their machines.

2. The author did not provide a proof

3. The second comment of the person who wrote the comment you quoted is this:
> Well I just rechecked Windows Media Player on my dad's pc, it forces LTR for control bar. Sorry I haven't been using Windows recently. But here's a screenshot for Rhythmbox on Linux.

There is no such thing as a physical media device with reversed controllers nor with text that is not in English.  It does not even make sense in business sense.

What would a manufacturer do if it does produce media devices with reversed controllers, and later finds out that there is too much of inventory which probably be delivered to another nation (not an oriental nation), would that manufacturer sell to Latin people media players with reversed controllers?

> I don't have enough information to decide either way, but I don't mind if an option is added for this.

No option. Just, literally, fixate layout to LTR.

**Comment by sjohannes** [comment]
It has to be an option, because according to [another comment in your link](http://sourceforge.net/p/smplayer/bugs/584/#80db), there *are* RTL locales in which the reversed layout is correct.

**Comment by sjohannes** [comment]
> Either way, this change is not a simple flip of a switch

Actually, sorry; I take that back. Your second link shows that it is that simple. OK, I'll make the change.

**Comment by GreenLunar** [comment]
> It has to be an option, because according to another comment in your link, there are RTL locales in which the reversed layout is correct.

> I don't need that. This is not a court of law. It's a reasonable assumption that you don't know the conventions in all RTL locales, so I tend to believe that both are valid, and thus an option is needed.

You are correct, I do not know them all, however as an Israeli whose native language is Hebrew, who knows Arabic and who also know Arabs in Egypt and Jordan, and who can assure that in Iran, even though I have never been there, there is the same practice implemented on physical devices, I **assume** that it is likewise in Pakistan and some couple more territories where people write from right-to-left.

IF there are places in GTK+ or Qt where media controllers are specifically and explicitly ordered from right-to-left it IS because the RTL-speaking people who developers have conversed with, did not put extra thought and just asked to reverse any thing.

You may like to bother these programmers in this concern `@srazi` (Iran, QupZilla web browser contributor) `@orgads` (Israel, Qt toolkit contributor).  I have asked them many requests, thus I feel uncomfortable to bother any of them.

**Comment by sjohannes** [comment]
Thanks, and sorry for not reading all the linked pages in your initial post before replying. This is fixed in the *gi* branch and appear in 4.0.

The playback controls are now always in this order: Previous, Play/Pause, Stop, Next, Seekbar. Button images, as well as the seekbar, are always in the LTR direction.

This can all be reversed by setting `gui/rtl_playback_controls`. This is not configurable from the Preferences UI and must be set from settings.ini. Depending on feedback, we can later add it to the UI or remove the option altogether.

**Comment by GreenLunar** [comment]
> Thanks, and sorry for not reading all the linked pages in your initial post before replying. This is fixed in the gi branch and will appear in 4.0.

You are welcome.

Please, do not sorry; this is a confusing matter.

Thank you very much for listening!

> This can all be reversed by setting `gui/rtl_playback_controls`. This is not configurable from the Preferences UI and must be set from settings.ini. Depending on feedback, we can later add it to the UI or remove the option altogether.

I hope it will be deprecated in future.  I will start notifying GNOME and KDE on this subject and hope to see them acting upon this issue.

**Comment by GreenLunar** [comment]
May you please post a screenshot?

I fail to run Exaile.

```
$ LANG=he_IL.utf8 exaile 
INFO    : Using PyGObject 3.10.1
WARNING : Exaile requires PyGObject 3.10.2 or greater!
INFO    : Loading Exaile 4.0.0dev on Python 2.7.5...
INFO    : Loading settings...
INFO    : Using he_IL UTF-8 locale
ERROR   : Unhandled exception
Traceback (most recent call last):
  File "/usr/lib/exaile/xl/main.py", line 169, in __init__
    self.__init()
  File "/usr/lib/exaile/xl/main.py", line 226, in __init
    from xlgui.widgets.info import Splash
  File "/usr/lib/exaile/xlgui/__init__.py", line 59, in <module>
    from xl import (
  File "/usr/lib/exaile/xl/player/__init__.py", line 46, in <module>
    PLAYER = player.ExailePlayer('player')
  File "/usr/lib/exaile/xl/player/player.py", line 79, in __init__
    self._setup_engine()
  File "/usr/lib/exaile/xl/player/player.py", line 91, in _setup_engine
    from .gst.engine import ExaileGstEngine
ImportError: No module named gst.engine
```

**Comment by sjohannes** [comment]
The *gi* branch uses GTK+ 3; see if your OS has a PyGObject version 3 package (e.g. python-gobject in Debian), I think that should pull all the required libraries.

Here is a screenshot with he_IL locale:

![exaile-rtl](https://cloud.githubusercontent.com/assets/8440927/10695642/13dacfb2-79e8-11e5-9ecf-b7ef20aacbdb.png)

**Comment by GreenLunar** [comment]
Thanks.  Exaile looks good!

I have later installed [pygobject3](http://slackbuilds.org/repository/14.1/python/pygobject3/), and even python3, gnome-common and py3cairo; still not working.

**Comment by GreenLunar** [comment]
P.S. I may help finish Hebrew translation of Exaile, if you would assure me that you would undo changes by users (mostly a person who wants to spread his contact details) I am not interested to have.

**Comment by virtuald** [comment]
@GreenLunar what error message are you getting? If it's the same one as above, you don't have pygobject installed correctly. It needs to be installed for python 2, as Exaile is not yet Python3 compatible.

We don't control translations particularly strictly -- in part because I have no idea what they say. :) All of our translations happen through the @weblate website -- if there are particular translations that are wrong, then feel free to change them. If there are bad translations by particular users, you might want to contact weblate about that.

**Comment by GreenLunar** [comment]
> It needs to be installed for python 2, as Exaile is not yet Python3 compatible.

I have removed `python3`, `gnome-common` and `py3cairo`, and recompiled `pygobject3`; now I get the previous error.

```
$ exaile 
INFO    : Using PyGObject 3.10.2
INFO    : Loading Exaile 4.0.0dev on Python 2.7.5...
INFO    : Loading settings...
INFO    : Using en_US UTF-8 locale
ERROR   : Unhandled exception
Traceback (most recent call last):
  File "/usr/lib/exaile/xl/main.py", line 169, in __init__
    self.__init()
  File "/usr/lib/exaile/xl/main.py", line 226, in __init
    from xlgui.widgets.info import Splash
  File "/usr/lib/exaile/xlgui/__init__.py", line 59, in <module>
    from xl import (
  File "/usr/lib/exaile/xl/player/__init__.py", line 46, in <module>
    PLAYER = player.ExailePlayer('player')
  File "/usr/lib/exaile/xl/player/player.py", line 79, in __init__
    self._setup_engine()
  File "/usr/lib/exaile/xl/player/player.py", line 91, in _setup_engine
    from .gst.engine import ExaileGstEngine
ImportError: No module named gst.engine
```

> if there are particular translations that are wrong, then feel free to change them.

I will ask you to revert commits of users I claim to be abusers.

> If there are bad translations by particular users, you might want to contact weblate about that.

I think `@nijel` already knows.

**Comment by virtuald** [comment]
If there are bad translations, you are welcome to change them via a pull request or via the weblate interface. If that isn't sufficient, we can deal with that on a case-by-case basis, just send me and/or @sjohannes an email. 

As far as your pygobject issue... that's rather strange. I suspect that your python is using the old pyc/pyo files present. Try removing them all, and running again. Something like this:

    find . -name '*.pyc' -delete
    find . -name '*.pyo' -delete

**Comment by GreenLunar** [comment]
> Try removing them all, and running again. Something like this:
> ```
> find . -name '*.pyc' -delete
> find . -name '*.pyo' -delete
> ```

I do not want to execute it from `/`.  May you please direct me to a specific directory?

**Comment by virtuald** [comment]
Oh. 

I bet the Makefile hasn't been updated yet. Ha. One sec...

**Comment by virtuald** [comment]
Do a `git pull`, and install it again. We hadn't updated the Makefile yet because we (well, at least I) typically just run exaile directly from git, without installing it. The `gi` branch hasn't been used very much by non-devs at this point, but hopefully we'll get out a release by the end of the year.

**Comment by GreenLunar** [comment]
Still not working:
```
$ exaile 
INFO    : Using PyGObject 3.10.2
INFO    : Loading Exaile 4.0.0dev on Python 2.7.5...
INFO    : Loading settings...
INFO    : Using en_US UTF-8 locale
ERROR   : Unhandled exception
Traceback (most recent call last):
  File "/usr/lib/exaile/xl/main.py", line 169, in __init__
    self.__init()
  File "/usr/lib/exaile/xl/main.py", line 226, in __init
    from xlgui.widgets.info import Splash
  File "/usr/lib/exaile/xlgui/__init__.py", line 59, in <module>
    from xl import (
  File "/usr/lib/exaile/xl/player/__init__.py", line 46, in <module>
    PLAYER = player.ExailePlayer('player')
  File "/usr/lib/exaile/xl/player/player.py", line 79, in __init__
    self._setup_engine()
  File "/usr/lib/exaile/xl/player/player.py", line 91, in _setup_engine
    from .gst.engine import ExaileGstEngine
  File "/usr/lib/exaile/xl/player/gst/__init__.py", line 19, in <module>
    logger.info("Using GStreamer %s.%s.%s", Gst.VERSION_MAJOR,
  File "/usr/lib/python2.7/site-packages/gi/module.py", line 313, in __getattr__
    return getattr(self._introspection_module, name)
  File "/usr/lib/python2.7/site-packages/gi/module.py", line 134, in __getattr__
    self.__name__, name))
AttributeError: 'gi.repository.Gst' object has no attribute 'VERSION_MAJOR'
```

Would you like to continue this discussion by email?

**Comment by virtuald** [comment]
Sure. You can find my email on my github profile page.
